### ISSUE TITLE
[JENKINS-28447] CpsScmFlowDefinition does not resolve variables

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/workflow-cps-plugin#1134]
<p>When selecting the "Groovy CPS DSL from SCM" option for a worflow job, the SCM plugins do not appear to resolve build parameters or environment variables.  I am using the git plugin and when I use it from other jobs I can specify a build parameter, like "BuildBranch", and use that when specifying what branch should be built:</p>

<p>Branches to build: */${BuildBranch}</p>

<p>This does not work when I use the Groovy CPS DSL from SCM.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nsnewland">nsnewland</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-28447" target="_blank">CpsScmFlowDefinition does not resolve variables</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Critical
<li><b>component(s)</b>: workflow-cps-plugin
<li><b>label(s)</b>: followup
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2017-04-26T14:09:28+00:00
<li><b>votes</b>: 50
<li><b>watchers</b>: 67
<li><b>imported</b>: 2025-12-07
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>When selecting the "Groovy CPS DSL from SCM" option for a worflow job, the SCM plugins do not appear to resolve build parameters or environment variables.  I am using the git plugin and when I use it from other jobs I can specify a build parameter, like "BuildBranch", and use that when specifying what branch should be built:</p>

<p>Branches to build: */${BuildBranch}</p>

<p>This does not work when I use the Groovy CPS DSL from SCM.</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Jenins 1.613, Workflow plugin 1.6<br/>
Also: Jenkins 1.651.1, workflow-job 2.3
```
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-28447?focusedId=239628&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-239628">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Well this would apply to any SCM plugin that has any field which gets variables substituted, I expect. Have not investigated in detail.</p>
</pre>
</details>

</ul>
</details>

<p>Well this would apply to any SCM plugin that has any field which gets variables substituted, I expect. Have not investigated in detail.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-28447?focusedId=251553&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-251553">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><tt>SCM</tt> implementations seem to call <tt>Run.getEnvironment(TaskListener)</tt> so it is possible an implementation of <a href="https://issue-redirect.jenkins.io/issue/30910" title="Job parameters should be exposed as environment variables" class="issue-link" data-issue-key="JENKINS-30910"><del>JENKINS-30910</del></a> would automatically fix this, but certainly deserves its own integration tests.</p>
</pre>
</details>

</ul>
</details>

<p><tt>SCM</tt> implementations seem to call <tt>Run.getEnvironment(TaskListener)</tt> so it is possible an implementation of <a href="https://issue-redirect.jenkins.io/issue/30910" title="Job parameters should be exposed as environment variables" class="issue-link" data-issue-key="JENKINS-30910"><del>JENKINS-30910</del></a> would automatically fix this, but certainly deserves its own integration tests.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kad.png" title="kad's avatar" /> kad:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-28447?focusedId=258303&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-258303">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>bump. without being  able to use variables in pipeline from scm is practically making pipeline type of job useless, as can't be integrated e.g. with Github pull requests processing.</p>
</pre>
</details>

</ul>
</details>

<p>bump. without being  able to use variables in pipeline from scm is practically making pipeline type of job useless, as can't be integrated e.g. with Github pull requests processing.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/docwhat.png" title="docwhat's avatar" /> docwhat:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-28447?focusedId=262015&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-262015">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>So this breaks pipeline jobs using Gerrit.  The Gerrit variables are not available in the "Pipeline script from SCM" section.  The are not replaced, but left as <tt>$FOOBAR</tt>; example: <tt>git -c core.askpass=true fetch --tags --progress ssh://jenkins@gerrit.example.com:29418/noodles/jenkinsfile.git $GERRIT_REFSPEC</tt></p>

<p>It breaks other cases too.  For example, in the same job above, I've set a build parameter <tt>$GERRIT_REFSPEC</tt> and it doesn't work either.</p>
</pre>
</details>

</ul>
</details>

<p>So this breaks pipeline jobs using Gerrit.  The Gerrit variables are not available in the "Pipeline script from SCM" section.  The are not replaced, but left as <tt>$FOOBAR</tt>; example: <tt>git -c core.askpass=true fetch --tags --progress ssh://jenkins@​gerrit.example.com:29418/noodles/jenkinsfile.git $GERRIT_REFSPEC</tt></p>

<p>It breaks other cases too.  For example, in the same job above, I've set a build parameter <tt>$GERRIT_REFSPEC</tt> and it doesn't work either.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>kmbulebu2:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-28447?focusedId=262251&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-262251">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The suggested workaround will not work if your script uses 'checkout scm'. </p>

<blockquote><p>{{java.lang.IllegalStateException: inappropriate context
	at org.jenkinsci.plugins.workflow.multibranch.SCMVar.getValue(SCMVar.java:75)}}</p></blockquote>

<p>'checkout scm' is only allowed for CpsScmFlowDefinition type jobs, not the CpsFlowDefinition type job used by the workaround.</p>

<p>In my case, I fully worked out a Pipeline script to build the master branch, then duped the job and modified it to trigger on patchset-created events in Gerrit. I suppose I'll go back to the original script and modify it to build its own scm object. I'll have to keep in mind that it could be invoked by either a ref-updated or patchset-created event, changing the parameters to the git scm class. It's a bit of a rabbit hole. </p>
</pre>
</details>

</ul>
</details>

<p>The suggested workaround will not work if your script uses 'checkout scm'. </p>

<blockquote><p>{{java.lang.IllegalStateException: inappropriate context<br/>
	at org.jenkinsci.plugins.workflow.multibranch.SCMVar.getValue(SCMVar.java:75)}}</p></blockquote>

<p>'checkout scm' is only allowed for CpsScmFlowDefinition type jobs, not the CpsFlowDefinition type job used by the workaround.</p>

<p>In my case, I fully worked out a Pipeline script to build the master branch, then duped the job and modified it to trigger on patchset-created events in Gerrit. I suppose I'll go back to the original script and modify it to build its own scm object. I'll have to keep in mind that it could be invoked by either a ref-updated or patchset-created event, changing the parameters to the git scm class. It's a bit of a rabbit hole. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/gareth_western.jpg" title="gareth_western's avatar" /> gareth_western:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-28447?focusedId=266054&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-266054">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is this a duplicate of <a href="https://issue-redirect.jenkins.io/issue/34876" title="Git Parameters not working for Pipeline projects and Jenkinsfile from SCM" class="issue-link" data-issue-key="JENKINS-34876">JENKINS-34876</a> and <a href="https://issue-redirect.jenkins.io/issue/34540" title="Can't use Pipeline Parameter in Branch Specifier on Jenkins 1.651.2/2.0-2.6" class="issue-link" data-issue-key="JENKINS-34540"><del>JENKINS-34540</del></a>?</p>
</pre>
</details>

</ul>
</details>

<p>Is this a duplicate of <a href="https://issue-redirect.jenkins.io/issue/34876" title="Git Parameters not working for Pipeline projects and Jenkinsfile from SCM" class="issue-link" data-issue-key="JENKINS-34876">JENKINS-34876</a> and <a href="https://issue-redirect.jenkins.io/issue/34540" title="Can't use Pipeline Parameter in Branch Specifier on Jenkins 1.651.2/2.0-2.6" class="issue-link" data-issue-key="JENKINS-34540"><del>JENKINS-34540</del></a>?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-28447?focusedId=267007&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-267007">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Have some evidence that <a href="https://issue-redirect.jenkins.io/issue/30910" title="Job parameters should be exposed as environment variables" class="issue-link" data-issue-key="JENKINS-30910"><del>JENKINS-30910</del></a> would fix this implicitly.</p>
</pre>
</details>

</ul>
</details>

<p>Have some evidence that <a href="https://issue-redirect.jenkins.io/issue/30910" title="Job parameters should be exposed as environment variables" class="issue-link" data-issue-key="JENKINS-30910"><del>JENKINS-30910</del></a> would fix this implicitly.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-28447?focusedId=270282&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-270282">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I think I have finally reproduced this in a test, the build ending with something like</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-none">…
 &gt; git rev-parse origin/${VERSION}^{commit} # timeout=10
 &gt; git rev-parse ${VERSION}^{commit} # timeout=10
ERROR: Couldn't find any revision to build. Verify the repository and branch configuration for this job.
Finished: FAILURE
</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p>I think I have finally reproduced this in a test, the build ending with something like</p>


<pre>
…
 > git rev-parse origin/${VERSION}^{commit} # timeout=10
 > git rev-parse ${VERSION}^{commit} # timeout=10
ERROR: Couldn't find any revision to build. Verify the repository and branch configuration for this job.
Finished: FAILURE
</pre>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jwillemsen.png" title="jwillemsen's avatar" /> jwillemsen:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-28447?focusedId=270697&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-270697">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Would be useful for our use case when the  Script Path could use the job name, that way we can have a generic setup where we just have to copy jobs and rename them, the script path will use the correct file for that job from scm</p>
</pre>
</details>

</ul>
</details>

<p>Would be useful for our use case when the  Script Path could use the job name, that way we can have a generic setup where we just have to copy jobs and rename them, the script path will use the correct file for that job from scm</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-28447?focusedId=270878&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-270878">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Jesse Glick
Path:
 pom.xml
 src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java
<a href="http://jenkins-ci.org/commit/workflow-cps-plugin/86c7ea988f371a8b50d2a2ed41ad2fcb4cd6de25" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/workflow-cps-plugin/86c7ea988f371a8b50d2a2ed41ad2fcb4cd6de25</a>
Log:
  <span class="error">&#91;FIXED JENKINS-28447&#93;</span> CpsScmFlowDefinition does not resolve variables.
Actual fix is implied by <a href="https://issue-redirect.jenkins.io/issue/30910" title="Job parameters should be exposed as environment variables" class="issue-link" data-issue-key="JENKINS-30910"><del>JENKINS-30910</del></a>; this is just the integration test.</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Jesse Glick<br/>
Path:<br/>
 pom.xml<br/>
 src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java<br/>
<a href="http://jenkins-ci.org/commit/workflow-cps-plugin/86c7ea988f371a8b50d2a2ed41ad2fcb4cd6de25" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/workflow-cps-plugin/86c7ea988f371a8b50d2a2ed41ad2fcb4cd6de25</a><br/>
Log:<br/>
  <span class="error">&#91;FIXED JENKINS-28447&#93;</span> CpsScmFlowDefinition does not resolve variables.<br/>
Actual fix is implied by <a href="https://issue-redirect.jenkins.io/issue/30910" title="Job parameters should be exposed as environment variables" class="issue-link" data-issue-key="JENKINS-30910"><del>JENKINS-30910</del></a>; this is just the integration test.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-28447?focusedId=270880&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-270880">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jwillemsen" class="user-hover" rel="jwillemsen">jwillemsen</a> this issue is about variable expansion in the SCM configuration. Your case is something else, best filed as a separate RFE. (Or really you may just want to be using the Job DSL plugin.)</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jwillemsen" class="user-hover" rel="jwillemsen">jwillemsen</a> this issue is about variable expansion in the SCM configuration. Your case is something else, best filed as a separate RFE. (Or really you may just want to be using the Job DSL plugin.)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>atikhono:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-28447?focusedId=297166&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-297166">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I don't understand why it's marked as resolved. I can still reproduce it with the latest workflow-cps (2.30) and workflow-job (2.10) plugins. Reproducible as it is in the description.</p>
<ul>
	<li>Create pipeline project</li>
	<li>In configure: check 'This project is parametrized' and add a String Parameter "branch" with some default value. Select 'Pipeline script from SCM' in definition. Fill in repository URL, Branch Specifier as ${branch} and Script Path (Jenkinsfile)</li>
	<li>At URL, create a git repository with Jenkinsfile. Put 'node('master') { checkout scm }' in the Jenkinsfile.</li>
	<li>Run the job. It will fail with: 
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">hudson.plugins.git.GitException: Command <span class="code-quote">"git fetch --tags --progress origin +refs/heads/${branch}:refs/remotes/origin/${branch} --prune"</span> returned status code 128:
stdout: 
stderr: fatal: Couldn't find remote ref refs/heads/${branch}

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1833)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1552)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$300(CliGitAPIImpl.java:66)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:343)
	at jenkins.plugins.git.GitSCMFileSystem$BuilderImpl.build(GitSCMFileSystem.java:306)
	at jenkins.scm.api.SCMFileSystem.of(SCMFileSystem.java:196)
	at jenkins.scm.api.SCMFileSystem.of(SCMFileSystem.java:172)
	at org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition.create(CpsScmFlowDefinition.java:99)
	at org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition.create(CpsScmFlowDefinition.java:59)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:232)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:405)
</pre>
</div></div></li>
</ul>

</pre>
</details>

</ul>
</details>

<p>I don't understand why it's marked as resolved. I can still reproduce it with the latest workflow-cps (2.30) and workflow-job (2.10) plugins. Reproducible as it is in the description.</p>
<ul>
	<li>Create pipeline project</li>
	<li>In configure: check 'This project is parametrized' and add a String Parameter "branch" with some default value. Select 'Pipeline script from SCM' in definition. Fill in repository URL, Branch Specifier as ${branch} and Script Path (Jenkinsfile)</li>
	<li>At URL, create a git repository with Jenkinsfile. Put 'node('master') { checkout scm }' in the Jenkinsfile.</li>
	<li>Run the job. It will fail with: 

<pre>
hudson.plugins.git.GitException: Command <span class="code-quote">"git fetch --tags --progress origin +refs/heads/${branch}:refs/remotes/origin/${branch} --prune"</span> returned status code 128:
stdout: 
stderr: fatal: Couldn't find remote ref refs/heads/${branch}

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1833)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1552)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$300(CliGitAPIImpl.java:66)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:343)
	at jenkins.plugins.git.GitSCMFileSystem$BuilderImpl.build(GitSCMFileSystem.java:306)
	at jenkins.scm.api.SCMFileSystem.of(SCMFileSystem.java:196)
	at jenkins.scm.api.SCMFileSystem.of(SCMFileSystem.java:172)
	at org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition.create(CpsScmFlowDefinition.java:99)
	at org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition.create(CpsScmFlowDefinition.java:59)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:232)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:405)
</pre></li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>atikhono:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-28447?focusedId=297170&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-297170">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Also, I'm facing the issue when configuring git reference repository in the Pipeline SCM section.</p>
<ul>
	<li>All the same configuration as in my comment above, but replace ${branch} with an existing branch, e.g master</li>
	<li>In 'Additional Behaviours' select 'Advanced Clone Behaviours'</li>
	<li>Set 'Path of the reference repo to use during clone' to some value depending on an environment variable, e.g. ${HOME}/jenkins/reference/big-git-repo.git. The job will report: 
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">Cloning repository ...URL...
 &gt; git init ... # timeout=10
ERROR: Reference path does not exist: ${HOME}/jenkins/reference/big-git-repo.git</pre>
</div></div></li>
</ul>


<p>Would be very handy to have this feature. In my case, I want to use the 'checkout scm' shortcut in my pipeline script that runs several builds in parallel on different platforms. So I want to specify where a reference repository exists on a specific node. For instance, on a Windows node I want to put it in D:\jenkins\reference, on a Mac – to /Users/jenkins/reference and, finally, on linux – /home/jenkins/reference. I could set an environment variable ${REFERENCE} to the above values in node configuration and than use it in 'Path of the reference repo to use during clone' as ${REFERENCE}.</p>
</pre>
</details>

</ul>
</details>

<p>Also, I'm facing the issue when configuring git reference repository in the Pipeline SCM section.</p>
<ul>
	<li>All the same configuration as in my comment above, but replace ${branch} with an existing branch, e.g master</li>
	<li>In 'Additional Behaviours' select 'Advanced Clone Behaviours'</li>
	<li>Set 'Path of the reference repo to use during clone' to some value depending on an environment variable, e.g. ${HOME}/jenkins/reference/big-git-repo.git. The job will report: 

<pre>
Cloning repository ...URL...
 > git init ... # timeout=10
ERROR: Reference path does not exist: ${HOME}/jenkins/reference/big-git-repo.git</pre></li>
</ul>


<p>Would be very handy to have this feature. In my case, I want to use the 'checkout scm' shortcut in my pipeline script that runs several builds in parallel on different platforms. So I want to specify where a reference repository exists on a specific node. For instance, on a Windows node I want to put it in D:\jenkins\reference, on a Mac – to /Users/jenkins/reference and, finally, on linux – /home/jenkins/reference. I could set an environment variable ${REFERENCE} to the above values in node configuration and than use it in 'Path of the reference repo to use during clone' as ${REFERENCE}.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-28447?focusedId=297182&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-297182">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=atikhono" class="user-hover" rel="atikhono">atikhono</a> you must disable lightweight checkout.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=atikhono" class="user-hover" rel="atikhono">atikhono</a> you must disable lightweight checkout.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>atikhono:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-28447?focusedId=297190&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-297190">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> Thanks, with lightweight checkout off ${branch} is resolved. Still, git reference repository path is not. Should I file another issue for that?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> Thanks, with lightweight checkout off ${branch} is resolved. Still, git reference repository path is not. Should I file another issue for that?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-28447?focusedId=297192&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-297192">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=atikhono" class="user-hover" rel="atikhono">atikhono</a> that sounds like an RFE for <tt>git-plugin</tt> not specific to Pipeline.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=atikhono" class="user-hover" rel="atikhono">atikhono</a> that sounds like an RFE for <tt>git-plugin</tt> not specific to Pipeline.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>asfaltboy:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-28447?focusedId=307540&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-307540">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> I came across this issue as well, and unchecking "lightweight checkout" also helped. However I cannot figure out why exactly, can you give a bit more context into why the env/parameter interpolation is disabled when LC is on ?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> I came across this issue as well, and unchecking "lightweight checkout" also helped. However I cannot figure out why exactly, can you give a bit more context into why the env/parameter interpolation is disabled when LC is on ?</p>
